Lib/ufo2ft/filters/propagateAnchors.py

```diff
@@ -0,0 +1,111 @@
+# Copyright 2016 Google Inc. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from fontTools.misc.transform import Transform
+from ufo2ft.filters import BaseFilter
+
+import logging
+
+
+logger = logging.getLogger(__name__)
+
+
+class PropagateAnchorsFilter(BaseFilter):
+
+    def set_context(self, font, glyphSet):
+        ctx = super(PropagateAnchorsFilter, self).set_context(font, glyphSet)
+        ctx.processed = set()
+        return ctx
+
+    def __call__(self, font, glyphSet=None):
+        if super(PropagateAnchorsFilter, self).__call__(font, glyphSet):
+            modified = self.context.modified
+            if modified:
+                logger.info('Glyphs with propagated anchors: %i' %
+                            len(modified))
+            return modified
+
+    def filter(self, glyph):
+        if not glyph.components:
+            return False
+        _propagate_glyph_anchors(glyph.font, glyph, self.context.processed)
+        return True
+
+
+def _propagate_glyph_anchors(ufo, parent, processed):
+    """Propagate anchors for a single parent glyph."""
+
+    if parent.name in processed:
+        return
+    processed.add(parent.name)
+
+    base_components = []
+    mark_components = []
+    anchor_names = set()
+    to_add = {}
+    for component in parent.components:
+        glyph = ufo[component.baseGlyph]
+        _propagate_glyph_anchors(ufo, glyph, processed)
+        if any(a.name.startswith('_') for a in glyph.anchors):
+            mark_components.append(component)
+        else:
+            base_components.append(component)
+            anchor_names |= {a.name for a in glyph.anchors}
+
+    for anchor_name in anchor_names:
+        # don't add if parent already contains this anchor OR any associated
+        # ligature anchors (e.g. "top_1, top_2" for "top")
+        if not any(a.name.startswith(anchor_name) for a in parent.anchors):
+            _get_anchor_data(to_add, ufo, base_components, anchor_name)
+
+    for component in mark_components:
+        _adjust_anchors(to_add, ufo, component)
+
+    # we sort propagated anchors to append in a deterministic order
+    for name, (x, y) in sorted(to_add.items()):
+        anchor_dict = {'name': name, 'x': x, 'y': y}
+        parent.appendAnchor(glyph.anchorClass(anchorDict=anchor_dict))
```

Comment:
it should be possible to just append a raw dict instead of using defcon specific API

Reply:
Actually neither will work with fontParts since it’s glyph `appendAnchor()` method takes the arguments `(name, position, color)`, whereas defcon’s takes an anchorDict dict or an Anchor object.

Reply:
Too bad.. well I still prefer a dict to that anchorClass